src/algorithms/statistical.py

```diff
@@ -9,26 +10,66 @@
 from nltk.probability import FreqDist
 import scipy.sparse
 
+from divergence import js_divergence
+
 
 def get_keyword(s, n=5):
     tok_sent = get_tokenized_sentences(s)
     terms, mat = get_coocurrence_matrix(tok_sent)
-    rated_terms = get_chi_squared(terms, mat)
+    clusters = cluster_terms(terms, mat)
+    rated_terms = get_chi_squared(terms, clusters, mat)
     sorted_terms = sorted(rated_terms.items(), key=operator.itemgetter(1))
     sorted_terms.reverse()
     keywords = sorted_terms[:n]
     return [i[0] for i in keywords]
 
 
-def get_chi_squared(terms, matrix):
+def get_chi_squared(terms, clusters, matrix):
     words = {}
     for term in terms:
-        prob = get_chi_squared_for_term(term, terms, matrix)
-        words.setdefault(term, prob)
+        words[term] = get_chi_squared_for_term(term, terms, clusters, matrix)
     return words
 
 
-def get_chi_squared_for_term(term, terms, matrix):
+def cluster_terms(terms, matrix):
+    threshold = 0.95 * log(2)
+    clusters = []
+
+    for term in terms:
+        dict_index = terms[term]
+        row = matrix.getrow(dict_index)
+        arr = row.toarray()[0]
+        arr[dict_index] = 0
+
+        cluster = set()
+        cluster.add(term)
+
+        for term2 in terms:
+            if term != term2:
+                dict_index = terms[term2]
+                row = matrix.getrow(dict_index)
+                arr2 = row.toarray()[0]
+                arr2[dict_index] = 0
+                div = js_divergence(arr, arr2)
+
+                if div > threshold:
+                    cluster.add(term2)
+
+        add_cluster(clusters, cluster)
+
+    return clusters
+
```

Comment:
I ran a 246 word abstract and it took about a half hour.  This function appears to be the culprit.  Any ideas on how to address this?